crates/wasm-conventions/src/lib.rs

```diff
@@ -148,3 +151,72 @@ pub fn get_or_insert_start_builder(module: &mut Module) -> &mut FunctionBuilder
         .unwrap_local_mut()
         .builder_mut()
 }
+
+pub fn insert_target_feature(module: &mut Module, new_feature: &str) -> Result<()> {
+    // Taken from <https://github.com/bytecodealliance/wasm-tools/blob/f1898f46bb9d96f0f09682415cb6ccfd6a4dca79/crates/wasmparser/src/limits.rs#L27>.
+    anyhow::ensure!(new_feature.len() <= 100_000, "feature name too long");
+
+    // Try to find an existing section.
+    let section = module
+        .customs
+        .iter_mut()
+        .find(|(_, custom)| custom.name() == "target_features");
+
+    // If one exists, check if the target feature is already present.
+    let section = if let Some((_, section)) = section {
+        let section: &mut RawCustomSection = section
+            .as_any_mut()
+            .downcast_mut()
+            .context("failed to read section")?;
+        let mut reader = BinaryReader::new(&section.data);
+        // The first integer contains the target feature count.
+        let count = reader.read_var_u32()?;
+
+        // Try to find if the target feature is already present.
+        for _ in 0..count {
+            // First byte is the prefix.
+            let prefix_index = reader.current_position();
+            let prefix = reader.read_u8()? as u8;
+            // Read the feature.
+            let length = reader.read_var_u32()?;
+            let feature = reader.read_bytes(length as usize)?;
+
+            // If we found the target feature, we are done here.
+            if feature == new_feature.as_bytes() {
+                // Make sure we set any existing prefix to "enabled".
+                if prefix == b'-' {
+                    section.data[prefix_index] = b'+';
+                }
+
+                return Ok(());
+            }
+        }
+
+        section
+    } else {
+        let mut data = Vec::new();
+        leb128::write::unsigned(&mut data, 0).unwrap();
+        let id = module.customs.add(RawCustomSection {
+            name: String::from("target_features"),
+            data,
+        });
+        module.customs.get_mut(id).unwrap()
+    };
+
+    // If we couldn't find the target feature, insert it.
+
+    // The first byte contains an integer describing the target feature count, which we increase by one.
+    let mut data = Cursor::new(&section.data);
+    let count = leb128::read::unsigned(&mut data).unwrap();
```

Comment:
This shouldn't use `unwrap`, since it can fail if `data` ends while it's still reading or the integer doesn't fit in a `u64`.

Reply:
This can't fail because we already read it successfully beforehand (or return an error on failure) or inserted it correctly ourselves if no section was found.

Reply:
Ah, good point, I didn't think about that.